mkosi/__init__.py

```diff
@@ -2218,21 +2218,68 @@ def finalize_initrds(context: Context) -> list[Path]:
     return [build_default_initrd(context)]
 
 
+def expand_kernel_specifiers(text: str, kver: str, token: str, roothash: str, boot_count: str) -> str:
+    ampersand = False
+    result: list[str] = []
+
+    specifiers = {
+        "&": "&",
+        "e": token,
+        "k": kver,
+        "h": roothash,
+        "c": boot_count
+    }
+
+    for c in text:
+        if ampersand:
+            ampersand = False
+
+            if specifier := specifiers.get(c):
+                result += specifier
+            else:
+                logging.warning(f"Unknown specifier '&{c}' found in {text}, ignoring")
+        elif c == "&":
+            ampersand = True
+        else:
+            result += c
+
+    if ampersand:
+        result += "&"
+
+    return "".join(result)
+
+
 def install_type1(
     context: Context,
     kver: str,
     kimg: Path,
     token: str,
     partitions: Sequence[Partition],
 ) -> None:
-    dst = context.root / "boot" / token / kver
-    entry = context.root / f"boot/loader/entries/{token}-{kver}.conf"
+    roothash = ""
+    if roothash_cli := finalize_roothash(partitions):
+        roothash = roothash_cli.partition("=")[2]
+
+    boot_count = ""
+    if (context.root / "etc/kernel/tries").exists():
+        boot_count = (context.root / "etc/kernel/tries").read_text().strip()
+
+    bootloader_entry = expand_kernel_specifiers(
+        context.config.bootloader_entry_format or "&e-&k",
+        kver=kver,
+        token=token,
+        roothash=roothash,
+        boot_count=boot_count,
+    )
+
+    dst = context.root / "boot" / bootloader_entry.replace("-", "/")
```

Comment:
Specifically this is going to cause problems for kver that have a `-` in the middle which I am relativly sure is very common.